libcontainer/state_linux.go

```diff
@@ -0,0 +1,206 @@
+	transition(containerState) error
+	destroy() error
+	status() Status
+}
+
+type destroyedState struct {
+	c *linuxContainer
+}
+
+func (b *destroyedState) status() Status {
+	return Destroyed
+}
+
+func (b *destroyedState) transition(s containerState) error {
+	switch s.(type) {
+	case *runningState:
+		b.c.state = s
+		return nil
+	case *restoredState:
+		b.c.state = s
+		return nil
+	case *destroyedState:
+		return nil
+	}
+	return newStateError(b, s)
+}
+
+func (b *destroyedState) destroy() error {
+	if !b.c.config.Namespaces.Contains(configs.NEWPID) {
+		if err := killCgroupProcesses(b.c.cgroupManager); err != nil {
+			logrus.Warn(err)
+		}
+	}
+	err := b.c.cgroupManager.Destroy()
+	if rerr := os.RemoveAll(b.c.root); err == nil {
+		err = rerr
+	}
+	b.c.initProcess = nil
+	if herr := b.runHooks(); err == nil {
+		err = herr
+	}
+	return err
+}
+
+func (b *destroyedState) runHooks() error {
+	if b.c.config.Hooks != nil {
+		s := configs.HookState{
+			Version: b.c.config.Version,
+			ID:      b.c.id,
+			Root:    b.c.config.Rootfs,
+		}
+		for _, hook := range b.c.config.Hooks.Poststop {
+			if err := hook.Run(s); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+type runningState struct {
+	base containerState
+	c    *linuxContainer
+}
+
+func (r *runningState) status() Status {
+	return Running
+}
+
+func (r *runningState) transition(s containerState) error {
+	switch s.(type) {
+	case *destroyedState:
+		running, err := r.c.isRunning()
+		if err != nil {
+			return err
+		}
+		if running {
+			return newGenericError(fmt.Errorf("container still running"), ContainerNotStopped)
+		}
+		r.c.state = s
+		return nil
+	case *pausedState:
+		r.c.state = s
+		return nil
+	case *runningState, *nullState:
+		return nil
+	}
+	return newStateError(r, s)
+}
+
+func (r *runningState) destroy() error {
+	running, err := r.c.isRunning()
+	if err != nil {
+		return err
+	}
+	if running {
+		return newGenericError(fmt.Errorf("container is not destroyed"), ContainerNotStopped)
+	}
+	return r.base.destroy()
+}
+
+type pausedState struct {
+	c *linuxContainer
+}
+
+func (p *pausedState) status() Status {
+	return Paused
+}
+
+func (p *pausedState) transition(s containerState) error {
+	switch s.(type) {
+	case *runningState:
+		p.c.state = s
+		return nil
+	case *pausedState:
+		return nil
+	}
+	return newStateError(p, s)
+}
+
+func (p *pausedState) destroy() error {
+	return newGenericError(fmt.Errorf("container is paused"), ContainerPaused)
+}
+
+type restoredState struct {
+	base     containerState
+	imageDir string
+	c        *linuxContainer
+}
+
+func (r *restoredState) status() Status {
+	return Running
+}
+
+func (r *restoredState) transition(s containerState) error {
+	switch s.(type) {
+	case *destroyedState:
+		return nil
+	case *runningState:
+		return nil
+	}
+	return newStateError(r, s)
+}
+
+func (r *restoredState) destroy() error {
+	if _, err := os.Stat(filepath.Join(r.c.root, "checkpoint")); err != nil {
```

Comment:
I believe this is correct.  That is not the checkpointed files but just a marker that the container was checkpointed.  It will be removed like all the state after the container stops.